kissim/comparison/tree.py

```diff
@@ -0,0 +1,246 @@
+    outputfile : str or pathlib.Path
+        Path to kinase tree file (TREE file) in Newick format.
+    cmethod : str
+        Clustering methods as offered by scipy, see [1]. Default is "ward", alternatives i.e.
+        "complete", "weighted", "average", "centroid".
+
+    Returns
+    -------
+    tree : scipy.cluster.hierarchy.ClusterNode
+        Cluster node.
+    mean_similarity : dict of int: float
+        Mean distance (value) for each node index (key).
+    """
+
+    input_path = Path(inputfile)
+    output_path = Path(outputfile)
+
+    # Read in KISSIM similarity matrix from provided inputfile
+    print(f"Reading kinase matrix from {inputfile}")
+    distance_matrix = pd.read_csv(inputfile, index_col=0)
+
+    # Generate tree
+    tree, mean_similarity = from_distance_matrix(distance_matrix, outputfile, clustering_method)
+
+    return tree, mean_similarity
+
+
+def from_distance_matrix(distance_matrix, outputfile=None, clustering_method="ward"):
+    """
+    Generate kissim-based kinase tree (cluster kinases and save clusters in the Newick format).
+
+    Parameters
+    ----------
+    distance_matrix : pandas.DataFrame
+        Distance matrix on which clustering is based.
+    outputfile : str or pathlib.Path
+        Path to kinase tree file (TREE file) in Newick format.
+    cmethod : str
+        Clustering methods as offered by scipy, see [1]. Default is "ward", alternatives i.e.
+        "complete", "weighted", "average", "centroid".
+
+    Returns
+    -------
+    tree : scipy.cluster.hierarchy.ClusterNode
+        Cluster node.
+    mean_similarity : dict of int: float
+        Mean distance (value) for each node index (key).
+
+    References
+    ----------
+    [1] https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html
+    """
+
+    outputfile = Path(outputfile)
+
+    # Removing problematic entries if they exist
+    for entry in PROBLAMATIC_KINASES:
+        if entry in distance_matrix:
+            distance_matrix.drop(entry, axis=0, inplace=True)
+            distance_matrix.drop(entry, axis=1, inplace=True)
+
+    # Curate diagonal - set to 0
+    np.fill_diagonal(distance_matrix.values, 0)
+
+    # Hierarchical clustering (Ward by default)
+    # Alternatives: 'complete', 'weighted', 'average', 'centroid'
+    print(f"Clustering (method: {clustering_method}) and calculating branch similarities")
+    hclust = sch.linkage(ssd.squareform(distance_matrix.values), method=clustering_method)
+    print("Converting clustering to a Newick tree")
+    tree = sch.to_tree(hclust, False)
+
+    # Calculate and assign mean similarity for each of the branches
+    mean_similarity = {}
+    _get_mean_index(tree, distance_matrix, mean_similarity)
+
+    # Save tree to file
+    if outputfile:
+        _to_newick(tree, mean_similarity, distance_matrix, outputfile)
+        _to_kinase_annotation(distance_matrix, outputfile.parent / "kinase_annotations.csv")
```

Comment:
It would be a **nice-to-have** to make this an optional routine that can be disabled with an argument.
This because this routine takes longer than the tree creation routine and will in practice (for most people) only be relevant on the first run or when changing set.

Reply:
Thanks for catching this!

If the user does not want to write out the tree and annotation files, they can set `outputfile=None`. 
I had to move the `outputfile` variable within the functions, so that no errors are thrown if `None`.

Reply:
I just realized you were only commenting on the annotation not the writing-files procedure in general. 

I am in favor of letting the user set both output paths themselves:
`inputfile` > `matrix_path` (must be set)
`outputfile` > `tree_path` + `annotation_path` (can both be `None` and no output file will be written)

Reply:
Even better!